File: modules/angular2/test/core/application_ref_spec.ts
import {
  ddescribe,
  describe,
  it,
  iit,
  xit,
  expect,
  beforeEach,
  afterEach,
  el,
  AsyncTestCompleter,
  fakeAsync,
  tick,
  inject,
  SpyObject
} from 'angular2/testing_internal';
import {Type} from 'angular2/src/facade/lang';
import {SpyChangeDetectorRef} from './spies';
import {
  ApplicationRef_,
  ApplicationRef,
  PLATFORM_CORE_PROVIDERS,
  APPLICATION_CORE_PROVIDERS
} from "angular2/src/core/application_ref";
import {
  Injector,
  Provider,
  APP_INITIALIZER,
  Component,
  ReflectiveInjector,
  coreLoadAndBootstrap,
  coreBootstrap,
  PlatformRef,
  createPlatform,
  disposePlatform,
  ComponentResolver,
  ChangeDetectorRef
} from "angular2/core";
import {Console} from 'angular2/src/core/console';
import {BaseException} from 'angular2/src/facade/exceptions';
import {PromiseWrapper, PromiseCompleter, TimerWrapper} from "angular2/src/facade/async";
import {
  ComponentFactory,
  ComponentRef_,
  ComponentRef
} from 'angular2/src/core/linker/component_factory';
import {ExceptionHandler} from 'angular2/src/facade/exception_handler';

export function main() {
  describe("bootstrap", () => {
    var platform: PlatformRef;
    var errorLogger: _ArrayLogger;
    var someCompFactory: ComponentFactory;

    beforeEach(() => {
      errorLogger = new _ArrayLogger();
      disposePlatform();
      platform = createPlatform(ReflectiveInjector.resolveAndCreate(PLATFORM_CORE_PROVIDERS));
      someCompFactory =
          new _MockComponentFactory(new _MockComponentRef(ReflectiveInjector.resolveAndCreate([])));
    });

    afterEach(() => { disposePlatform(); });

    function createApplication(providers: any[]): ApplicationRef_ {
      var appInjector = ReflectiveInjector.resolveAndCreate(
          [
            APPLICATION_CORE_PROVIDERS,
            new Provider(Console, {useValue: new _MockConsole()}),
            new Provider(ExceptionHandler, {useValue: new ExceptionHandler(errorLogger, false)}),
            new Provider(ComponentResolver,
                         {useValue: new _MockComponentResolver(someCompFactory)}),
            providers
          ],
          platform.injector);
      return appInjector.get(ApplicationRef);
    }

    describe("ApplicationRef", () => {
      it("should throw when reentering tick", () => {
        var cdRef = <any>new SpyChangeDetectorRef();
        var ref = createApplication([]);
        try {
          ref.registerChangeDetector(cdRef);
          cdRef.spy("detectChanges").andCallFake(() => ref.tick());
          expect(() => ref.tick()).toThrowError("ApplicationRef.tick is called recursively");
        } finally {
          ref.unregisterChangeDetector(cdRef);
        }
      });

      describe('run', () => {
        it('should rethrow errors even if the exceptionHandler is not rethrowing', () => {
          var ref = createApplication([]);
          expect(() => ref.run(() => { throw new BaseException('Test'); })).toThrowError('Test');
        });

        it('should return a promise with rejected errors even if the exceptionHandler is not rethrowing',
           inject([AsyncTestCompleter, Injector], (async, injector) => {
             var ref = createApplication([]);
             var promise = ref.run(() => PromiseWrapper.reject('Test', null));
             PromiseWrapper.catchError(promise, (e) => {
               expect(e).toEqual('Test');
               async.done();
             });
           }));
      });
    });

    describe("coreLoadAndBootstrap", () => {
      it("should wait for asynchronous app initializers",
         inject([AsyncTestCompleter, Injector], (async, injector) => {
           let completer: PromiseCompleter<any> = PromiseWrapper.completer();
           var initializerDone = false;
           TimerWrapper.setTimeout(() => {
             completer.resolve(true);
             initializerDone = true;
           }, 1);
           var app = createApplication(
               [new Provider(APP_INITIALIZER, {useValue: () => completer.promise, multi: true})]);
           coreLoadAndBootstrap(app.injector, MyComp)
               .then((compRef) => {
                 expect(initializerDone).toBe(true);
                 async.done();
               });
         }));
    });

    describe("coreBootstrap", () => {
      it("should throw if an APP_INITIIALIZER is not yet resolved",
         inject([Injector], (injector) => {
           var app = createApplication([
             new Provider(APP_INITIALIZER,
                          {useValue: () => PromiseWrapper.completer().promise, multi: true})
           ]);
           expect(() => app.bootstrap(someCompFactory))
               .toThrowError(
                   "Cannot bootstrap as there are still asynchronous initializers running. Wait for them using waitForAsyncInitializers().");
         }));
    });
  });
}

@Component({selector: 'my-comp', template: ''})
class MyComp {
}

class _ArrayLogger {
  res: any[] = [];
  log(s: any): void { this.res.push(s); }
  logError(s: any): void { this.res.push(s); }
  logGroup(s: any): void { this.res.push(s); }
  logGroupEnd(){};
}

class _MockComponentFactory extends ComponentFactory {
  constructor(private _compRef: ComponentRef) { super(null, null, null); }
  create(injector: Injector, projectableNodes: any[][] = null,
         rootSelectorOrNode: string | any = null): ComponentRef {
    return this._compRef;
  }
}

class _MockComponentResolver implements ComponentResolver {
  constructor(private _compFactory: ComponentFactory) {}

  resolveComponent(type: Type): Promise<ComponentFactory> {
    return PromiseWrapper.resolve(this._compFactory);
  }
  clearCache() {}
}

class _MockComponentRef extends ComponentRef_ {
  constructor(private _injector: Injector) { super(null, null); }
  get injector(): Injector { return this._injector; }
  get changeDetectorRef(): ChangeDetectorRef { return <any>new SpyChangeDetectorRef(); }
  onDestroy(cb: Function) {}
}

class _MockConsole implements Console {
  log(message) {}
}

File: test/core/application_ref_spec.dart
library angular2.test.core.application_ref_spec;

import "package:angular2/testing_internal.dart"
    show
        ddescribe,
        describe,
        it,
        iit,
        xit,
        expect,
        beforeEach,
        afterEach,
        el,
        AsyncTestCompleter,
        fakeAsync,
        tick,
        inject,
        SpyObject;
import "spies.dart" show SpyChangeDetector;
import "package:angular2/src/core/application_ref.dart"
    show ApplicationRef_, ApplicationRef, PlatformRef_;
import "package:angular2/core.dart" show Injector, Provider, APP_INITIALIZER;
import "package:angular2/src/core/change_detection/change_detector_ref.dart"
    show ChangeDetectorRef_;
import "package:angular2/src/facade/async.dart"
    show PromiseWrapper, PromiseCompleter, TimerWrapper;
import "package:angular2/src/facade/collection.dart" show ListWrapper;

main() {
  describe("ApplicationRef", () {
    it("should throw when reentering tick", () {
      var cd = (new SpyChangeDetector() as dynamic);
      var ref = new ApplicationRef_(null, null, null);
      ref.registerChangeDetector(new ChangeDetectorRef_(cd));
      cd.spy("detectChanges").andCallFake(() => ref.tick());
      expect(() => ref.tick())
          .toThrowError("ApplicationRef.tick is called recursively");
    });
  });
  describe("PlatformRef", () {
    describe("asyncApplication", () {
      void expectProviders(Injector injector, List<dynamic> providers) {
        for (var i = 0; i < providers.length; i++) {
          var provider = providers[i];
          expect(injector.get(provider.token)).toBe(provider.useValue);
        }
      }
      it(
          "should merge syncronous and asyncronous providers",
          inject([AsyncTestCompleter, Injector], (async, injector) {
            var ref = new PlatformRef_(injector, null);
            var ASYNC_PROVIDERS = [new Provider(Foo, useValue: new Foo())];
            var SYNC_PROVIDERS = [new Provider(Bar, useValue: new Bar())];
            ref
                .asyncApplication(
                    (zone) => PromiseWrapper.resolve(ASYNC_PROVIDERS),
                    SYNC_PROVIDERS)
                .then((appRef) {
              var providers =
                  ListWrapper.concat(ASYNC_PROVIDERS, SYNC_PROVIDERS);
              expectProviders(appRef.injector, providers);
              async.done();
            });
          }));
      it(
          "should allow function to be null",
          inject([AsyncTestCompleter, Injector], (async, injector) {
            var ref = new PlatformRef_(injector, null);
            var SYNC_PROVIDERS = [new Provider(Bar, useValue: new Bar())];
            ref.asyncApplication(null, SYNC_PROVIDERS).then((appRef) {
              expectProviders(appRef.injector, SYNC_PROVIDERS);
              async.done();
            });
          }));
      mockAsyncAppInitializer(completer,
          [List<dynamic> providers = null, Injector injector]) {
        return () {
          if (providers != null) {
            expectProviders(injector, providers);
          }
          TimerWrapper.setTimeout(() => completer.resolve(true), 1);
          return completer.promise;
        };
      }
      SpyObject createSpyPromiseCompleter() {
        var completer = PromiseWrapper.completer();
        var completerSpy = (new SpyObject() as dynamic);
        // Note that in TypeScript we need to provide a value for the promise attribute

        // whereas in dart we need to override the promise getter
        completerSpy.promise = completer.promise;
        completerSpy.spy("get:promise").andReturn(completer.promise);
        completerSpy.spy("resolve").andCallFake(completer.resolve);
        completerSpy.spy("reject").andCallFake(completer.reject);
        return completerSpy;
      }
      it(
          "should wait for asyncronous app initializers",
          inject([AsyncTestCompleter, Injector], (async, injector) {
            var ref = new PlatformRef_(injector, null);
            var completer = createSpyPromiseCompleter();
            var SYNC_PROVIDERS = [
              new Provider(Bar, useValue: new Bar()),
              new Provider(APP_INITIALIZER,
                  useValue: mockAsyncAppInitializer(completer), multi: true)
            ];
            ref.asyncApplication(null, SYNC_PROVIDERS).then((appRef) {
              expectProviders(
                  appRef.injector,
                  ListWrapper.slice(
                      SYNC_PROVIDERS, 0, SYNC_PROVIDERS.length - 1));
              expect(completer.spy("resolve")).toHaveBeenCalled();
              async.done();
            });
          }));
      it(
          "should wait for async providers and then async app initializers",
          inject([AsyncTestCompleter, Injector], (async, injector) {
            var ref = new PlatformRef_(injector, null);
            var ASYNC_PROVIDERS = [new Provider(Foo, useValue: new Foo())];
            var completer = createSpyPromiseCompleter();
            var SYNC_PROVIDERS = [
              new Provider(Bar, useValue: new Bar()),
              new Provider(APP_INITIALIZER,
                  useFactory: (injector) => mockAsyncAppInitializer(
                      completer, ASYNC_PROVIDERS, injector),
                  multi: true,
                  deps: [Injector])
            ];
            ref
                .asyncApplication(
                    (zone) => PromiseWrapper.resolve(ASYNC_PROVIDERS),
                    SYNC_PROVIDERS)
                .then((appRef) {
              expectProviders(
                  appRef.injector,
                  ListWrapper.slice(
                      SYNC_PROVIDERS, 0, SYNC_PROVIDERS.length - 1));
              expect(completer.spy("resolve")).toHaveBeenCalled();
              async.done();
            });
          }));
    });
    describe("application", () {
      it(
          "should throw if an APP_INITIIALIZER returns a promise",
          inject([Injector], (injector) {
            var ref = new PlatformRef_(injector, null);
            var appInitializer = new Provider(APP_INITIALIZER,
                useValue: () => PromiseWrapper.resolve([]), multi: true);
            expect(() => ref.application([appInitializer])).toThrowError(
                "Cannot use asyncronous app initializers with application. Use asyncApplication instead.");
          }));
    });
  });
}

class Foo {
  Foo() {}
}

class Bar {
  Bar() {}
}
